Thanks for the contribution, @chr1syy — appreciate the effort to tighten the YAML contract and document the agent-authored shape, this is exactly the kind of round-trip robustness Cue needs.

A couple of items before we can merge:

**1. CI lint is failing on the new validator code (blocker).**

The `lint-and-format` job is red on TypeScript errors introduced by this PR (not the unrelated `@types/*` issues you noted in the description):

```
src/main/cue/config/cue-config-validator.ts(374,8):  error TS18046: 'sub.source_session' is of type 'unknown'.
src/main/cue/config/cue-config-validator.ts(374,38): error TS18046: 'sub.source_sub' is of type 'unknown'.
src/main/cue/config/cue-config-validator.ts(376,41): error TS18046: 'sub.source_sub' is of type 'unknown'.
src/main/cue/config/cue-config-validator.ts(376,103): error TS18046: 'sub.source_session' is of type 'unknown'.
```

The issue is that storing `Array.isArray(sub.source_session)` in a local `const` doesn't carry the type-narrowing back to `sub.source_session` itself, so subsequent `.length` accesses are still on `unknown`. Inlining the `Array.isArray()` checks in the `if` conditions (or narrowing into local variables typed as `unknown[]`) will let TS narrow correctly.

Could you run `npm run lint` locally and push the fix? That should also surface anything else hidden by the `--no-verify` push.

**2. Missing symmetric validation case (nice-to-have, but Greptile flagged it as a real bug).**

Both CodeRabbit and Greptile noticed that the new array-alignment block doesn't cover scalar `source_session` + array `source_sub`. Greptile's read is that this is more than a nit: in `yamlToPipeline.ts` the loop iterates `sourcePositions.length` (= 1 for a scalar session), so any extra `source_sub` entries past index 0 get silently dropped — defeating the positional-alignment guarantee this PR is trying to enforce. Worth adding the fourth-shape error and a regression test:

```ts
} else if (!sourceSessionIsArray && sourceSubIsArray) {
    errors.push(`${prefix}: "source_sub" must be a string when "source_session" is a string`);
}
```

**3. Could you rebase onto latest `rc` once the above are in?**

`rc` has moved on a bit since this branch was cut and a clean rebase will give us a tidier merge.

Once CI is green and the symmetric case is covered I'm happy to take another look.

--- a/src/__tests__/main/cue/cue-yaml-loader.test.ts
+++ b/src/__tests__/main/cue/cue-yaml-loader.test.ts
@@ -795,6 +795,28 @@ subscriptions:
 			);
 		});
 
+		it('rejects source_sub array when source_session is a string', () => {
+			const result = validateCueConfig({
+				subscriptions: [
+					{
+						name: 'fan-in-invalid-shape',
+						event: 'agent.completed',
+						source_session: 'A',
+						source_sub: ['chain-a', 'chain-b'],
+						prompt: '{{CUE_SOURCE_OUTPUT}}',
+					},
+				],
+			});
+			expect(result.valid).toBe(false);
+			expect(result.errors).toEqual(
+				expect.arrayContaining([
+					expect.stringContaining(
+						'"source_sub" must be a string when "source_session" is a string'
+					),
+				])
+			);
+		});
+
 		it('accepts prompt_file as alternative to prompt', () => {
 			const result = validateCueConfig({
 				subscriptions: [

--- a/src/main/cue/config/cue-config-validator.ts
+++ b/src/main/cue/config/cue-config-validator.ts
@@ -368,16 +368,20 @@ function validateEventSpecificFields(
 		}
 		// For fan-in chains, source_sub should align positionally with
 		// source_session so each upstream source maps to its exact upstream sub.
-		const sourceSessionIsArray = Array.isArray(sub.source_session);
-		const sourceSubIsArray = Array.isArray(sub.source_sub);
-		if (sourceSessionIsArray && sourceSubIsArray) {
-			if (sub.source_session.length !== sub.source_sub.length) {
+		const sourceSession = sub.source_session;
+		const sourceSub = sub.source_sub;
+		const sourceSessionIsArray = Array.isArray(sourceSession);
+		const sourceSubIsArray = Array.isArray(sourceSub);
+		if (Array.isArray(sourceSession) && Array.isArray(sourceSub)) {
+			if (sourceSession.length !== sourceSub.length) {
 				errors.push(
-					`${prefix}: "source_sub" length (${sub.source_sub.length}) must match "source_session" length (${sub.source_session.length})`
+					`${prefix}: "source_sub" length (${sourceSub.length}) must match "source_session" length (${sourceSession.length})`
 				);
 			}
-		} else if (sourceSessionIsArray && typeof sub.source_sub === 'string') {
+		} else if (sourceSessionIsArray && typeof sourceSub === 'string') {
 			errors.push(`${prefix}: "source_sub" must be an array when "source_session" is an array`);
+		} else if (!sourceSessionIsArray && sourceSubIsArray) {
+			errors.push(`${prefix}: "source_sub" must be a string when "source_session" is a string`);
 		}
 	} else if (event === 'task.pending') {
 		if (!sub.watch || typeof sub.watch !== 'string') {

--- a/src/renderer/components/CuePipelineEditor/utils/yamlToPipeline.ts
+++ b/src/renderer/components/CuePipelineEditor/utils/yamlToPipeline.ts
@@ -532,6 +532,7 @@ export function subscriptionsToPipelines(
 			if (aIdx !== bIdx) return aIdx - bIdx;
 			return a.name.localeCompare(b.name);
 		});
+		const knownSubNames = new Set(sorted.map((s) => s.name));
 
 		let triggerCount = 0;
 		let columnIndex = 0;
@@ -862,7 +863,11 @@ export function subscriptionsToPipelines(
 						const bySubRef = subRef ? subNameToNode.get(subRef) : undefined;
 						if (bySubRef) {
 							sourceNode = bySubRef;
-						} else if (typeof subRef === 'string' && subRef.length > 0) {
+						} else if (
+							typeof subRef === 'string' &&
+							subRef.length > 0 &&
+							!knownSubNames.has(subRef)
+						) {
 							const errorNodeId = `error-source-sub-${sub.name}-${i}`;
 							sourceNode = createErrorNode(
 								errorNodeId,
@@ -953,7 +958,11 @@ export function subscriptionsToPipelines(
 					const bySubRef = subRef ? subNameToNode.get(subRef) : undefined;
 					if (bySubRef) {
 						resolvedSources.push(bySubRef);
-					} else if (typeof subRef === 'string' && subRef.length > 0) {
+					} else if (
+						typeof subRef === 'string' &&
+						subRef.length > 0 &&
+						!knownSubNames.has(subRef)
+					) {
 						const errorNodeId = `error-source-sub-${sub.name}-${i}`;
 						const errorNode = createErrorNode(
 							errorNodeId,

